bolt-modules/boltlib/lib/puppet/functions/run_task_with.rb

```diff
@@ -0,0 +1,195 @@
+  # @param block A block that returns a `Hash` of target-specific parameters for the task.
+  # @return A list of results, one entry per target.
+  # @example Run a task with target-specific parameters and a description
+  #   run_task_with('my_task', $targets, 'Update system packages') |$t| {
+  #     { 'param1' => $t.vars['var1'],
+  #       'param2' => $t.vars['var2'] }
+  #   }
+  dispatch :run_task_with_with_description do
+    param 'String[1]', :task_name
+    param 'Boltlib::TargetSpec', :targets
+    param 'Optional[String]', :description
+    optional_param 'Hash[String[1], Any]', :options
+    required_block_param 'Callable[Target]', :block
+    return_type 'ResultSet'
+  end
+
+  def run_task_with(task_name, targets, options = {}, &block)
+    run_task_with_with_description(task_name, targets, nil, options, &block)
+  end
+
+  def run_task_with_with_description(task_name, targets, description, options = {})
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue
+        .from_issue_and_stack(
+          Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING,
+          action: 'run_task_with'
+        )
+    end
+
+    executor  = Puppet.lookup(:bolt_executor)
+    inventory = Puppet.lookup(:bolt_inventory)
+    error_set = []
+
+    # Report to analytics
+    executor.report_function_call(self.class.name)
+    executor.report_bundled_content('Task', task_name)
+
+    # Keep valid metaparameters, discarding everything else
+    options = options.select { |k, _v| k.start_with?('_') }
+                     .transform_keys { |k| k.sub(/^_/, '').to_sym }
+
+    options[:description] = description if description
+
+    # Get all the targets
+    targets = Array(inventory.get_targets(targets))
+
+    # If all targets use the 'pcp' transport, use a fake task instead of loading the local definition
+    # Otherwise, load the local task definition
+    if (pcp_only = targets.any? && targets.all? { |t| t.transport == 'pcp' })
+      task = Bolt::Task.new(task_name, {}, [{ 'name' => '', 'path' => '' }])
+    else
+      task_signature = Puppet::Pal::ScriptCompiler.new(closure_scope.compiler).task_signature(task_name)
+
+      if task_signature.nil?
+        raise Bolt::Error.unknown_task(task_name)
+      end
+
+      task = Bolt::Task.from_task_signature(task_signature)
+    end
+
+    # Map the targets to their specific parameters and merge with the defaults
+    target_mapping = targets.each_with_object({}) do |target, mapping|
+      params = yield(target)
+
+      # Parameters returned from the block should be a Hash. If they're not, create a failing
+      # Result for the target that will later be added to the ResultSet.
+      unless params.is_a?(Hash)
+        exception = with_stack(
+          :TYPE_MISMATCH,
+          "Block must return a Hash of parameters, received #{params.class}"
+        )
+        error_set << Bolt::Result.from_exception(target, exception, action: 'task')
```

Comment:
Hm, I think the common case here is going to be that users have defined a hash where the values are target attributes. If this error is raised it's likely that there's a typo in the block somewhere, and all targets will fail. I wonder if it's better to fail the plan here so that it doesn't need to check the output for every target just to fail all of them. Mostly I think the case where this fails for only a few targets will be rare, and that the output will be annoying for more than a handful of targets and will waste cycles:
```
lucy ~/githubs/bolt ((8add2fa...)) $ be bolt plan run test -t docker
Starting: plan test
Starting: install puppet and gather facts on 127.0.0.1:20022, 127.0.0.1:20023, 127.0.0.1:20024
Finished: install puppet and gather facts with 0 failures in 3.4 sec
Starting: description on 
Finished: description with 0 failures in 0.0 sec
Finished: plan test in 3.41 sec
Failed on 127.0.0.1:20022:
  Block must return a Hash of parameters, received String (file: /home/lucy/githubs/bolt/site-modules/test/plans/init.pp, line: 6)
Failed on 127.0.0.1:20023:
  Block must return a Hash of parameters, received String (file: /home/lucy/githubs/bolt/site-modules/test/plans/init.pp, line: 6)
Failed on 127.0.0.1:20024:
  Block must return a Hash of parameters, received String (file: /home/lucy/githubs/bolt/site-modules/test/plans/init.pp, line: 6)
Failed on 3 targets: 127.0.0.1:20022,127.0.0.1:20023,127.0.0.1:20024
Ran on 3 targets
```